### ISSUE TITLE
noqa that affects lints on other lines: E402

### ISSUE THREAD

**Issue by asottile** [PyCQA/pycodestyle#880]
:wave: 

I'm scoping implementing: warn on unused `noqa` comments (for flake8) and I've noticed a problem with my initial approach, the E402 error can be suppressed by a `noqa` comment on a line that is not the line where the error occurs

## current state

Take the following:

```python
import os

x = 1

import foo
```

which produces:

```console
$ ~/opt/venv/bin/pycodestyle t.py
t.py:5:1: E402 module level import not at top of file
```

there are two ways currently to suppress this:

```python
import os

x = 1  # noqa

import foo
```

and

```python
import os

x = 1

import foo  # noqa
```

===============

## proposal 1

the first proposal would be to change the reported location of the errors to the non-import offending line of code.  in files where there are lots of imports after some code (take for instance `gevent` monkepatching) the entirety of the (many) E402 would be silence-able with a single `noqa` (still) but the reported line and noqa line would be consistent such that the "warn unused noqas" can work

## proposal 2

don't allow `noqa` on the non-import offending line of code and require it to be on the individual imports afterwards

this has a negative in that those which are using the first noqa strategy above would have to add many new noqa comments on each import, but it would satisfy what I want to do

## proposal 3

???

===================

related: https://github.com/asottile/yesqa/issues/19  https://github.com/PyCQA/pycodestyle/issues/871

**Comment by ederag** [comment]
Has this ship sailed
(after a flake8 upgrade, there were plenty of `E402` no longer silenced by a `# noqa: E402` on the non-import line,
so that looks like proposal 2, and that would be confirmed by https://github.com/PyCQA/flake8/issues/739#issuecomment-812821723),
or is left open for discussion ?

**Comment by asottile** [comment]
yeah pycodestyle is wrong here and shouldn't suppress errors when noqa isn't on the line if the error

**Comment by ederag** [comment]
OK,  I'll move the code lines after the imports or ignore E402.
The following is just food for thought (I'm just a user while you have overview and insight).

----

Prior to 3.8 the `# noqa: E402` was about ignoring the line, as far as `E402` is concerned,
and that made more sense to me.
(`noqa` meaning "exclude from quality assessment", not "suppress error")
So I do not see the former `pycodestyle` as "wrong" here.

But indeed that yields https://github.com/asottile/yesqa/issues/19,
which could also be solved by ignoring E402 in [yesqa](https://github.com/asottile/yesqa),
because it is too hard to solve fully
(who is the culprit: the import line after code or the code line before import ?),
short of trying without `# noqa: E402`, but that would require a flake8 pass.
